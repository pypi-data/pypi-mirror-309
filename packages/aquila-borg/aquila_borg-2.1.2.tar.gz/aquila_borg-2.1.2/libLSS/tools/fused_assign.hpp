/*+
    ARES/HADES/BORG Package -- ./libLSS/tools/fused_assign.hpp
    Copyright (C) 2014-2020 Guilhem Lavaux <guilhem.lavaux@iap.fr>
    Copyright (C) 2009-2020 Jens Jasche <jens.jasche@fysik.su.se>

    Additional contributions from:
       Guilhem Lavaux <guilhem.lavaux@iap.fr> (2023)
    
+*/
#ifndef __LIBLSS_FUSED_ASSIGNMENT_HPP
#define __LIBLSS_FUSED_ASSIGNMENT_HPP

#include <iostream>
#include <boost/type_traits/has_trivial_constructor.hpp>
#include "libLSS/tools/console.hpp"

// When we can get rid of AssignFunctor
//#include "libLSS/tools/phoenix_vars.hpp"
//#include <boost/phoenix/operator.hpp>

namespace LibLSS {

    namespace FUSE_details {
      template<std::size_t N, typename BiFunctor, bool parallel>
      struct OperatorAssignment {};

      template<std::size_t N, typename BiFunctor>
      struct OperatorAssignment<N,BiFunctor,false> {
        template<typename A, typename B>
        static inline void apply(BiFunctor f, A&& a, const B& b) {
          std::size_t s = a.index_bases()[0], e = a.shape()[0];
          for (std::size_t i = s; i < s+e; i++) {
            OperatorAssignment<N-1,BiFunctor,false> op;
            op.apply(f, a[i], b[i]);
          }
        }
      };


      template<typename BiFunctor>
      struct OperatorAssignment<3,BiFunctor,false> {
        template<typename A, typename B>
        static inline void apply(BiFunctor f, A&& a, const B& b) {
          auto ib = a.index_bases();
          auto sh = a.shape();
          std::size_t s0 = ib[0], e0 = s0+sh[0];
          std::size_t s1 = ib[1], e1 = s1+sh[1];
          std::size_t s2 = ib[2], e2 = s2+sh[2];
          boost::array<ssize_t, 3> i;
          for (i[0] = s0; i[0] < e0; i[0]++) {
            for (i[1] = s1; i[1] < e1; i[1]++) {
              for (i[2] = s2; i[2] < e2; i[2]++) {
                f(a(i), b(i));
              }
            }
          }
        }
      };

      template<std::size_t N, typename BiFunctor>
      struct OperatorAssignment<N,BiFunctor,true> {
        template<typename A, typename B>
        static inline void apply(BiFunctor f, A&& a, const B& b) {
          std::size_t s = a.index_bases()[0], e = a.shape()[0];
          typename boost::remove_reference<A>::type *a_ptr = &a;
          const B *b_ptr = &b;

#pragma omp parallel for schedule(static)
          for (std::size_t i = s; i < s+e; i++) {
            OperatorAssignment<N-1,BiFunctor,false> op;
            op.apply(f, (*a_ptr)[i], (*b_ptr)[i]);
          }
        }
      };


      template<typename BiFunctor>
      struct OperatorAssignment<3,BiFunctor,true> {
        template<typename A, typename B>
        static inline void apply(BiFunctor f, A&& a, const B& b) {
          auto ib = a.index_bases();
          auto sh = a.shape();
          std::size_t s0 = ib[0], e0 = s0+sh[0];
          std::size_t s1 = ib[1], e1 = s1+sh[1];
          std::size_t s2 = ib[2], e2 = s2+sh[2];
//          Console::instance().print<LOG_DEBUG>("Using optimized 3-loop collapsed omp");
          Console::instance().format<LOG_DEBUG>("Using optimized 3-loop collapsed omp, %dx%dx%d -- %dx%dx%d", s0,s1,s2,e0,e1,e2);

#pragma omp parallel for collapse(3)
          for (size_t i = s0; i < e0; i++) {
            for (size_t j = s1; j < e1; j++) {
              {
/*                boost::array<ssize_t, 3> idx;
                idx[0] = i;
                idx[1] = j;

                for (idx[2] = s2; idx[2] < e2; idx[2]++) {
                  f( a(idx), b(idx));
                }
		*/
		      for (size_t k = s2; k < e2; k++) {
			      f(a[i][j][k], b[i][j][k]);
		      }
            }
          }
        }
      }
      };

      template<typename BiFunctor>
      struct OperatorAssignment<2,BiFunctor,true> {
        template<typename A, typename B>
        static inline void apply(BiFunctor f, A&& a, const B& b) {
          std::size_t s0 = a.index_bases()[0], e0 = a.shape()[0];
          std::size_t s1 = a.index_bases()[1], e1 = a.shape()[1];

#pragma omp parallel for collapse(1) schedule(static)
          for (std::size_t i = s0; i < s0+e0; i++) {
            // Factorize memory access for the last index.
            auto stripe_a = a[i];
            auto stripe_b = b[i];
            for (std::size_t j = s1; j < s1+e1; j++) {
              f( a[i][j], b[i][j]);
            }
          }
        }
      };

      template<typename BiFunctor>
      struct OperatorAssignment<0, BiFunctor, false> {
        template<typename A, typename B>
        static inline void apply(BiFunctor f, A& a, const B& b) {
          f(a,b);
        }
      };

      template<typename BiFunctor>
      struct OperatorAssignment<0, BiFunctor, true> {
        template<typename A, typename B>
        static inline void apply(BiFunctor f, A& a, const B& b) {
          f(a,b);
        }
      };

      struct AssignFunctor {
        template<typename T0,typename T1>
        inline void operator()(T0& a, const T1& b) {
          a = b;
        }
      };


      template<typename OutArray, typename BiFunctor, typename InArray>
      inline void apply_array(BiFunctor f, OutArray A, const InArray& B, bool openmp = true) {
        typedef typename boost::remove_reference<OutArray>::type PureArray;
        if (openmp) {
          OperatorAssignment<PureArray::dimensionality,BiFunctor,true> op;
          op.template apply(f, A, B);
        } else {
          OperatorAssignment<PureArray::dimensionality,BiFunctor,false> op;
          op.template apply(f, A, B);
        }
      }

      template<typename OutArray, typename InArray>
      inline void copy_array_rv(OutArray A, const InArray& B, bool openmp=true) {
        // GCC is not yet sufficiently clever for that one. The code is suboptimal (test_fused_array timing)
        //        auto assigner = (boost::phoenix::ref(_p1)=boost::phoenix::cref(_p2));
        AssignFunctor assigner;
        apply_array<OutArray, decltype(assigner), InArray>(assigner, A, B, openmp);
      }

      template<typename OutArray, typename InArray>
      inline void copy_array(OutArray& A, const InArray& B, bool openmp=true) {
         copy_array_rv<OutArray&>(A, B, openmp);
      }

    }

    using FUSE_details::apply_array;
    using FUSE_details::copy_array;
    using FUSE_details::copy_array_rv;

};

#endif
