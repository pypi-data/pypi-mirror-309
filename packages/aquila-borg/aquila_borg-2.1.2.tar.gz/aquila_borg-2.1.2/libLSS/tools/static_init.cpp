/*+
    ARES/HADES/BORG Package -- ./libLSS/tools/static_init.cpp
    Copyright (C) 2014-2020 Guilhem Lavaux <guilhem.lavaux@iap.fr>
    Copyright (C) 2009-2020 Jens Jasche <jens.jasche@fysik.su.se>

    Additional contributions from:
       Guilhem Lavaux <guilhem.lavaux@iap.fr> (2023)
    
+*/
#include "libLSS/tools/static_init.hpp"

LibLSS::StaticInit& LibLSS::StaticInit::instance() {
    static StaticInit singleton;
        
    return singleton;
}
