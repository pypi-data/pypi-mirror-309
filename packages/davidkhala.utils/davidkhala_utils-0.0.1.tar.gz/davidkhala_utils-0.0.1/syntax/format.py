import json


def JSONReadable(data):
    if type(data) != dict:
        data = json.loads(data)
    return json.dumps(data, indent=4, sort_keys=True)
