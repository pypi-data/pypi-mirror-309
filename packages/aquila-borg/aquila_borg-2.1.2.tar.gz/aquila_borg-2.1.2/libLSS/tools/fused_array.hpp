/*+
    ARES/HADES/BORG Package -- ./libLSS/tools/fused_array.hpp
    Copyright (C) 2014-2020 Guilhem Lavaux <guilhem.lavaux@iap.fr>
    Copyright (C) 2009-2020 Jens Jasche <jens.jasche@fysik.su.se>

    Additional contributions from:
       Guilhem Lavaux <guilhem.lavaux@iap.fr> (2023)
    
+*/
#ifndef __LIBLSS_FUSED_ARRAY_HPP
#define __LIBLSS_FUSED_ARRAY_HPP

#include <boost/multi_array.hpp>

#include "libLSS/tools/nary_arrays.hpp"
#include "libLSS/tools/fused_assign.hpp"
#include "libLSS/tools/ref_tools.hpp"

namespace LibLSS {

  namespace FUSE_detail {

    namespace FusedArrayTypes {

      typedef boost::multi_array_types::size_type size_type;
      typedef boost::multi_array_types::index index;

    } // namespace FusedArrayTypes

    template <typename T, std::size_t ND, typename Allocator>
    struct BoostArray {
      enum { NumDims = ND };
      typedef boost::multi_array<T, ND, Allocator> type;
      typedef T element;
    };

    template <typename T, std::size_t ND>
    struct BoostRefArray {
      enum { NumDims = ND };
      typedef boost::const_multi_array_ref<T, ND> type;
      typedef T element;
    };

    template <typename T>
    struct ConstantArray {
      enum { NumDims = 0 };
      typedef T type;
      typedef T element;
    };

    struct FusedArray_base {
      typedef FusedArrayTypes::size_type size_type;
      typedef FusedArrayTypes::index index;
    };

    template <
        typename ArrayTuple, typename Operation, std::size_t N,
        std::size_t order, bool shaped>
    struct FusedArray_view;

    template <typename ArrayTuple, typename Operation, std::size_t N>
    struct FusedArray_view_scalar : FusedArray_base {
      static constexpr bool Shaped = ArrayTuple::Shaped;
      Operation op;
      const ArrayTuple &tuple;
      typedef typename ArrayTuple::element element;
      typedef FusedArrayTypes::size_type size_type;
      typedef boost::array<index, N> subindex;
      subindex idx;

      inline FusedArray_view_scalar(
          Operation _op, const ArrayTuple &_tuple, const subindex &i)
          : op(_op), tuple(_tuple), idx(i) {}

      inline const element &apply() const {
        return details::apply_op(op, tuple, idx);
      };

      inline operator element() const {
        return details::apply_op(op, tuple, idx);
      }
    };

    // *************************** Begin of view with no shape support ****************

    template <
        typename ArrayTuple, typename Operation, std::size_t N,
        std::size_t order>
    struct FusedArray_view<ArrayTuple, Operation, N, order, false>
        : FusedArray_base {
      static constexpr bool Shaped = ArrayTuple::Shaped;
      Operation op;
      ArrayTuple tuple;

      typedef boost::array<index, N> subindex;
      typedef FusedArray_view<ArrayTuple, Operation, N, order - 1, false>
          subview;
      typedef FusedArray_view<ArrayTuple, Operation, N, 0, false> subview0;
      subindex idx;

      inline FusedArray_view(
          Operation _op, const ArrayTuple &_tuple, const subindex &i)
          : op(_op), tuple(_tuple), idx(i) {}

      template <typename T_sub>
      inline typename std::enable_if<
          subindex::static_size == T_sub::static_size, subview0>::type
      operator()(T_sub const &i) const {
        return subview0(op, tuple, i);
      }

      template <typename T_sub>
      inline typename std::enable_if<
          order != N &&
              (N - order + T_sub::static_size == subindex::static_size),
          subview0>::type
      operator()(T_sub const &i) const {
        subindex j;
        for (unsigned int k = 0; k < N - order; k++)
          j[k] = idx[k];
        for (unsigned int k = N - order; k < N; k++)
          j[k] = i[k - (N - order)];

        return subview0(op, tuple, j);
      }

      inline subview operator[](size_type i) const {
        subindex idx0 = idx;
        idx0[N - order] = i;
        return subview(op, tuple, idx0);
      }
    };

    template <typename ArrayTuple, typename Operation, std::size_t N>
    struct FusedArray_view<ArrayTuple, Operation, N, 0, false>
        : FusedArray_view_scalar<ArrayTuple, Operation, N> {
      static constexpr bool Shaped = ArrayTuple::Shaped;
      typedef FusedArray_view_scalar<ArrayTuple, Operation, N> super;
      typedef typename super::subindex subindex;

      inline FusedArray_view(
          Operation _op, const ArrayTuple &_tuple, const subindex &i)
          : FusedArray_view_scalar<ArrayTuple, Operation, N>(_op, _tuple, i) {}
    };

    // a (0,0) is never shaped.
    template <typename ArrayTuple, typename Operation>
    struct FusedArray_view<ArrayTuple, Operation, 0, 0, false>
        : FusedArray_base {
      static constexpr bool Shaped = ArrayTuple::Shaped;
      typedef typename ArrayTuple::element element;
      Operation op;
      const ArrayTuple &tuple;
      const index *indexes;
      const size_type *shapes;
      typedef boost::array<size_type, 0> subindex;
      typedef FusedArray_view<ArrayTuple, Operation, 0, 0, false> subview;

      inline FusedArray_view(
          Operation _op, const ArrayTuple &_tuple, const subindex &i)
          : op(_op), tuple(_tuple) {}

      inline const element &apply() const {
        return details::apply_op(op, tuple, subindex());
      };

      inline operator element() const {
        return details::apply_op(op, tuple, subindex());
      }

      inline subview operator[](size_type i) const {
        // Ignore everything
        return subview(op, tuple, subindex());
      }
    };

    // *************************** End of view with no shape support ****************

    // *************************** Begin of view with shape support ****************

    template <
        typename ArrayTuple, typename Operation, std::size_t N,
        std::size_t order>
    struct FusedArray_view<ArrayTuple, Operation, N, order, true>
        : FusedArray_base {
      typedef typename ArrayTuple::element element;
      Operation op;
      ArrayTuple tuple;

      static constexpr bool Shaped = ArrayTuple::Shaped;
      typedef boost::array<index, N> subindex_root;
      typedef boost::array<index, order> subindex;
      typedef FusedArray_view<ArrayTuple, Operation, N, order - 1, true>
          subview;
      typedef FusedArray_view<ArrayTuple, Operation, N, 0, true> subview0;
      subindex_root idx;
      const index *v_index;
      const size_type *v_size;

      inline FusedArray_view(
          Operation _op, const ArrayTuple &_tuple, const subindex_root &i,
          const index *_index, const size_type *_size)
          : op(_op), tuple(_tuple), idx(i), v_index(_index), v_size(_size) {}

      template <typename T = element>
      inline typename std::enable_if<order == 1, T>::type
      operator[](size_type i) const {
        subindex_root idx0 = idx;
        idx0[N - order] = i;
        return details::apply_op(op, tuple, idx0);
      }

      template <typename T = subview>
      inline typename std::enable_if<order != 1, T>::type
      operator[](size_type i) const {
        subindex_root idx0 = idx;
        idx0[N - order] = i;
        return subview(op, tuple, idx0, v_index + 1, v_size + 1);
      }

      inline subview0 operator()(subindex_root const &i) const {
        return subview0(op, tuple, i, 0, 0);
      }

      template <typename T = subview0>
      inline typename std::enable_if<order != N, T>::type
      operator()(subindex const &i) const {
        subindex_root idx0 = idx;
        for (unsigned int k = 0; k < order; k++)
          idx0[N - order + k] = i[k];
        return subview0(op, tuple, idx0, 0, 0);
      }

      inline const size_type *shape() const { return v_size; }

      inline const index *index_bases() const { return v_index; }
    };

    template <typename ArrayTuple, typename Operation, std::size_t N>
    struct FusedArray_view<ArrayTuple, Operation, N, 0, true>
        : FusedArray_view_scalar<ArrayTuple, Operation, N> {
      typedef FusedArray_view_scalar<ArrayTuple, Operation, N> super;
      static constexpr bool Shaped = ArrayTuple::Shaped;
      typedef typename super::subindex subindex;
      typedef typename super::index index;
      typedef typename super::size_type size_type;

      inline FusedArray_view(
          Operation _op, const ArrayTuple &_tuple, const subindex &i,
          const index *_index, const size_type *_size)
          : FusedArray_view_scalar<ArrayTuple, Operation, N>(_op, _tuple, i) {}
      // Here we discard _index and _size as we have just a zero-dim array.
    };

    // *************************** End of view with shape support ****************

    template <std::size_t N>
    struct NumDimDecrement {
      enum { value = N - 1 };
    };

    template <>
    struct NumDimDecrement<0> {
      enum { value = 0 };
    };

    template <
        typename ArrayTuple, typename Operation,
        bool ShapeTuple = ArrayTuple::Shaped>
    struct FusedArray;

    template <typename ArrayTuple, typename Operation>
    struct FusedArray<ArrayTuple, Operation, true> : FusedArray_base {
      enum { NumDims = ArrayTuple::NumDims };
      enum { dimensionality = ArrayTuple::NumDims };
      typedef typename ArrayTuple::element element;
      static constexpr bool Shaped = true;

      ArrayTuple const a;
      Operation const op;
      typedef boost::array<index, NumDims> subindex;
      typedef FusedArray_view<
          ArrayTuple, Operation, NumDims, NumDimDecrement<NumDims>::value,
          ArrayTuple::Shaped>
          subview;

      inline FusedArray(ArrayTuple p_a, Operation p_op) : a(p_a), op(p_op) {}

      inline subview operator[](FusedArrayTypes::size_type i) const {
        subindex idx;
        if (NumDims >= 1)
          idx[0] = i;
        return subview(op, a, idx, a.index_bases() + 1, a.shape() + 1);
      }

      inline element operator()(const subindex &idx) const {
        return details::apply_op(op, a, idx);
      }

      inline const size_type *shape() const { return a.shape(); }

      inline const index *index_bases() const { return a.index_bases(); }

      inline size_type num_elements() const { return a.num_elements(); }
    };

    // No shape then strip all the shape related functions.
    template <typename ArrayTuple, typename Operation>
    struct FusedArray<ArrayTuple, Operation, false> : FusedArray_base {
      enum { NumDims = ArrayTuple::NumDims };
      enum { dimensionality = ArrayTuple::NumDims };
      typedef typename ArrayTuple::element element;
      static constexpr bool Shaped = false;

      ArrayTuple const a;
      Operation const op;
      typedef boost::array<index, NumDims> subindex;
      typedef FusedArray_view<
          ArrayTuple, Operation, NumDims, NumDimDecrement<NumDims>::value,
          ArrayTuple::Shaped>
          subview;

      inline FusedArray(ArrayTuple p_a, Operation p_op) : a(p_a), op(p_op) {}

      inline subview operator[](FusedArrayTypes::size_type i) const {
        subindex idx;
        if (NumDims >= 1)
          idx[0] = i;
        return subview(op, a, idx);
      }

      inline element operator()(const subindex &idx) const {
        return details::apply_op(op, a, idx);
      }
    };

    // This is a new API using Variadic templates. The interface is incompatible with b_fused
    // thus a new name. The operator has to be put at the front of the list to
    // support an arbitrary number of arrays in the list.
    // This variant can only be used if there is at least one array
    // in the list, thus this specialization.
    // The no-array variant is as usual using the ArrayNullTuple variant.

    // This function is compatible both with temporary objects and stable objects.
    // The universal reference mechanism (Array1&& etc) grabs everything and produce
    // references, even to rvalues. However reference to rvalues is unsafe and likely produce
    // segv in most cases for our specific use. So we strip the rvalue reference and convert
    // it back to a normal copy in that case. That allows slicing, intricated fused calls etc.

    template <typename A>
    using remove_const_ref = typename boost::remove_const<
        typename boost::remove_reference<A>::type>::type;

    template <
        typename Return, typename Operation, typename Array1,
        typename... Arrays>
    inline auto b_va_fused(Operation op, Array1 &&a1, Arrays &&... arrays) {
      typedef typename boost::remove_reference<Array1>::type BareArray1;
      typedef std::tuple<
          typename strip_rvalue_ref<decltype(a1)>::type,
          typename strip_rvalue_ref<decltype(arrays)>::type...>
          btuple;
      typedef ArrayTuple<
          BareArray1::dimensionality, Return, btuple,
          DetectShaped<typename boost::remove_const<BareArray1>::type>::Shaped>
          RealTuple;

      return FusedArray<RealTuple, Operation>(
          RealTuple(btuple(a1, arrays...)), op);
    }

    template <typename Return, std::size_t N, typename Operation>
    inline auto b_va_fused(Operation op) {
      typedef ArrayNullTuple<N, Return, N> RealTuple;
      return FusedArray<RealTuple, Operation>(RealTuple(), op);
    }

    template <
        typename Return, typename Array1, typename Array2, typename Array3,
        typename Array4, typename Operation>
    inline auto b_fused(
        const Array1 &a1, const Array2 &a2, const Array3 &a3, const Array4 &a4,
        Operation op) {
      typedef typename std::tuple<
          const Array1 &, const Array2 &, const Array3 &, const Array4 &>
          tuple;
      typedef ArrayTuple<Array1::dimensionality, Return, tuple, true> RealTuple;

      return FusedArray<RealTuple, Operation>(
          RealTuple(std::forward_as_tuple(a1, a2, a3, a4)), op);
    }

    template <
        typename Return, typename Array1, typename Array2, typename Array3,
        typename Operation>
    inline auto b_fused(
        const Array1 &a1, const Array2 &a2, const Array3 &a3, Operation op) {
      typedef
          typename std::tuple<const Array1 &, const Array2 &, const Array3 &>
              tuple;
      typedef ArrayTuple<Array1::dimensionality, Return, tuple, true> RealTuple;

      return FusedArray<RealTuple, Operation>(
          RealTuple(std::forward_as_tuple(a1, a2, a3)), op);
    }

    template <
        typename Return, typename Array1, typename Array2, typename Operation>
    inline auto b_fused(const Array1 &a1, const Array2 &a2, Operation op) {
      typedef typename std::tuple<const Array1 &, const Array2 &> tuple;
      typedef ArrayTuple<Array1::dimensionality, Return, tuple, true> RealTuple;

      return FusedArray<RealTuple, Operation>(
          RealTuple(std::forward_as_tuple(a1, a2)), op);
    }

    template <typename Return, typename Array1, typename Operation>
    inline auto b_fused(const Array1 &a1, Operation op) {
      typedef typename std::tuple<const Array1 &> tuple;
      typedef ArrayTuple<Array1::dimensionality, Return, tuple, true> RealTuple;
      return FusedArray<RealTuple, Operation>(
          RealTuple(std::forward_as_tuple(a1)), op);
    }

    template <typename Return, std::size_t N, typename Operation>
    inline auto b_fused(Operation op) {
      typedef ArrayNullTuple<N, Return> RealTuple;
      return FusedArray<RealTuple, Operation>(RealTuple(), op);
    }

    /**
       * This builds a virtual array with N-dimensions, whose elements
       * are of type Return. Each access to the element call the provided
       * operator, with the corresponding requested index. 
       * The signature should thus be "op(size_t, ...)->Return" with number of arguments
       * corresponding to the number of dimensions.
       * This array is "infinite", i.e. it does not have any specified bound, which may be
       * required for some assignment operation.
       *
       * @param op a functor that must be Copy-Constructible
       * @return a Copy-Constructible expression
       */
    template <typename Return, std::size_t N, typename Operation>
    inline auto b_fused_idx(Operation op) {
      typedef ArrayNullTuple<N, Return, N> RealTuple;
      return FusedArray<RealTuple, Operation>(RealTuple(), op);
    }

    /**
       * This builds a virtual array with N-dimensions, whose elements
       * are of type Return. Each access to the element call the provided
       * operator, with the corresponding requested index. 
       * The signature should thus be "op(size_t, ...)->Return" with number of arguments
       * corresponding to the number of dimensions.
       * This array is finite with extents indicated by "f_extents".
       *
       * @param op a functor that must be Copy-Constructible.
       * @param f_extents extent of the new virtual array.
       * @return a Copy-Constructible expression
       */
    template <
        typename Return, std::size_t N, typename ExtentType, typename Operation>
    inline auto b_fused_idx(Operation op, const ExtentType &f_extents) {
      typedef ArrayNullTupleExtent<N, Return, N> RealTuple;
      return FusedArray<RealTuple, Operation>(RealTuple(f_extents), op);
    }

    /**
       * Same as the above, but with extents split into two lists for easier management.
       */
    template <
        typename Return, std::size_t N, typename ShapeList, typename IndexList,
        typename Operation>
    inline auto b_fused_idx(
        Operation op, const ShapeList &_shapes, const IndexList &_indexes) {
      typedef ArrayNullTupleExtent<N, Return, N> RealTuple;
      return FusedArray<RealTuple, Operation>(RealTuple(_shapes, _indexes), op);
    }

  } // namespace FUSE_detail

  using FUSE_detail::b_fused;
  using FUSE_detail::b_fused_idx;
  using FUSE_detail::b_va_fused;

} // namespace LibLSS

#endif
