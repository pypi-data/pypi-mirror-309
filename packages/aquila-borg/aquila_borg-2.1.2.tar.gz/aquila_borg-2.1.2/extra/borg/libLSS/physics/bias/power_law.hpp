/*+
    ARES/HADES/BORG Package -- ./extra/borg/libLSS/physics/bias/power_law.hpp
    Copyright (C) 2014-2020 Guilhem Lavaux <guilhem.lavaux@iap.fr>
    Copyright (C) 2009-2020 Jens Jasche <jens.jasche@fysik.su.se>

    Additional contributions from:
       Guilhem Lavaux <guilhem.lavaux@iap.fr> (2023)
    
+*/
#ifndef __LIBLSS_PHYSICS_POWER_LAW_BIAS_HPP
#define __LIBLSS_PHYSICS_POWER_LAW_BIAS_HPP

// This header provides the implementations of a simple power law bias model.
// A generic bias model must implement the following concept:
//     a "selection_adaptor" functor, available from the object
//     a "compute_density(final_density)" function accepting a virtual array and returning a tuple of virtual array.
//     a "apply_adjoint_gradient(final_density,gradient_likelihood_array)" also accepting virtual arrays and returning a virtual array
//     a "prepare" function which is called before the other two calls, to allow for precomputing auxiliary complex fields.
//     a "cleanup" which allows for tearing down any auxiliary fields and memory allocated in the prepare phase.

#include <cmath>
#include <boost/bind.hpp>
#include "libLSS/tools/fused_array.hpp"
#include <tuple>
#include "libLSS/tools/phoenix_vars.hpp"
#include <boost/phoenix/operator.hpp>
#include <boost/phoenix/stl/cmath.hpp>
#include "libLSS/tools/tuple_helper.hpp"
#include "libLSS/physics/bias/base.hpp"

namespace LibLSS {

  namespace bias {

    namespace detail {

      inline double dummy_lambda(double) { return 0; }

      struct PowerLaw {
        static constexpr double EPSILON_VOIDS = 1e-6;
        static constexpr int numParams = 2;
        static constexpr const bool NmeanIsBias = true;

        double alpha;
        double nmean;

        // This adaptor transforms an unselected galaxy density (with appropriate
        // auxiliary arrays) to a selected array. It must be a functor accepting two
        // parameters: a selection virtual array and a bias density virtual array.
        selection::SimpleAdaptor selection_adaptor;

        PowerLaw(LikelihoodInfo const& = LikelihoodInfo()) {}

        template <typename BiasParameters>
        static inline void setup_default(BiasParameters &params) {
          params[0] = 10;
          params[1] = 0.2;
        }

        // Nota: fwd_model and final_density arrays cannot be stored in this step. But
        // they can be used.
        template <
            class ForwardModel, typename FinalDensityArray,
            typename BiasParameters, typename MetaSelect = NoSelector>
        inline void prepare(
            ForwardModel &fwd_model, const FinalDensityArray &final_density,
            double const _nmean, const BiasParameters &params,
            bool density_updated, MetaSelect _select = MetaSelect()) {
          // No need for complex preparation. Just copy parameters.
          nmean = params[0];
          alpha = params[1];
        }

        inline void cleanup() {
          // No need for cleanup
        }

        static inline double gradient_density_lambda(
            double nmean, double alpha, double v, double g) {
          return alpha * nmean * std::pow(1 + v + EPSILON_VOIDS, alpha - 1) * g;
        }

        static inline double
        density_lambda(double nmean, double alpha, double v) {
          double v0 = nmean * std::pow(1 + EPSILON_VOIDS + v, alpha);
          Console::instance().c_assert(!std::isinf(v0), "V is infinite");
          return v0;
        }

        inline double get_linear_bias() const { return alpha; }

        template <typename Array>
        static inline bool check_bias_constraints(Array &&a) {
          return a[0] > 0 && a[1] > 0 && a[1] < 5; // alpha can take any value
        }

        // This function returns an array-like array. That array
        // depends on the existence of the final density array.
        // The return type is quite complex. Let the compiler decides.
        // C++11 does not allow automatic return type deduction. C++14 would be
        // needed for that. So we have to rely on an auxiliary function that
        // allow for a compact decltype to be written.
        template <typename FinalDensityArray>
        inline auto compute_density(const FinalDensityArray &array) {
          return std::make_tuple(b_va_fused<double>(
              std::bind(
                  density_lambda, nmean, alpha, std::placeholders::_1),
              array));
        }

        // This function returns an array-like array. That array
        // depends on the existence of the final density array and the gradient likelihood array.
        // That is the job of the caller to ensure that temporary variables are not cleared
        // before the final use.
        // The return type is quite complex. Let the compiler decides.
        // L(b_0(delta, p), b_1(delta, p), ..., b_n(delta, p))
        // Now we take a tuple of gradient and collapse this to a gradient of delta.
        template <
            typename FinalDensityArray, typename TupleGradientLikelihoodArray>
        inline auto apply_adjoint_gradient(
            const FinalDensityArray &array,
            TupleGradientLikelihoodArray grad_array) {
          BOOST_STATIC_ASSERT(
              (std::tuple_size<TupleGradientLikelihoodArray>::value == 1));
          return std::make_tuple(b_va_fused<double>(
              std::bind(
                  gradient_density_lambda, nmean, alpha,
                  std::placeholders::_1, std::placeholders::_2),
              array, std::move(std::get<0>(grad_array))));
        }
      };

    } // namespace detail

    using detail::PowerLaw;

  } // namespace bias

} // namespace LibLSS

#endif
