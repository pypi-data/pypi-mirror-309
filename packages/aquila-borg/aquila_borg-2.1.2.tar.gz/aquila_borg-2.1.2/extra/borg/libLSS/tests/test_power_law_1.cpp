/*+
    ARES/HADES/BORG Package -- ./extra/borg/libLSS/tests/test_power_law_1.cpp
    Copyright (C) 2014-2020 Guilhem Lavaux <guilhem.lavaux@iap.fr>
    Copyright (C) 2009-2020 Jens Jasche <jens.jasche@fysik.su.se>

    Additional contributions from:
       Guilhem Lavaux <guilhem.lavaux@iap.fr> (2023)
    
+*/
#define BOOST_TEST_MODULE power_law_1
#define BOOST_TEST_NO_MAIN
#define BOOST_TEST_ALTERNATIVE_INIT_API

#include <boost/test/included/unit_test.hpp>
#include <boost/test/data/test_case.hpp>

#include <iostream>
#include <boost/format.hpp>
#include <boost/multi_array.hpp>
#include "libLSS/physics/bias/power_law.hpp"
#include "libLSS/tools/fused_array.hpp"
#include "libLSS/tools/fused_assign.hpp"
#include "libLSS/tools/static_init.hpp"

namespace utf = boost::unit_test;
using boost::extents;
using boost::format;
using LibLSS::bias::PowerLaw;

struct Model {
} model;
size_t N = 32;
auto d_extents = extents[N][N][N];
boost::array<double, 2> params{1, 2.0};
boost::multi_array<double, 3> final_density(d_extents);
boost::multi_array<double, 3> grad_density(d_extents);
PowerLaw law;
static constexpr auto EPSILON_VOIDS = PowerLaw::EPSILON_VOIDS;

BOOST_AUTO_TEST_CASE(warmup) {
  law.prepare(model, final_density, 1, params, true);

  // Variant using phoenix lambda capabilities
  LibLSS::copy_array(
      final_density, LibLSS::b_fused_idx<double, 3>(
                         [](size_t i, size_t j, size_t k) -> size_t {
                           return ((1 + i) * (1 + j) * (1 + k));
                         }));

  // Variant with C++ lambda, could be emulated with phoenix but quite artificial
  LibLSS::copy_array(
      grad_density, LibLSS::b_fused_idx<double, 3>(
                        [](int i, int j, int k) -> double { return 1; }));
}

BOOST_AUTO_TEST_CASE(shape_test, *utf::depends_on("warmup")) {
  // Derive the galaxy density
  auto density_g = std::get<0>(law.compute_density(final_density));

  BOOST_CHECK(density_g.shape()[0] == N);
  BOOST_CHECK(density_g.shape()[1] == N);
  BOOST_CHECK(density_g.shape()[2] == N);
}

BOOST_AUTO_TEST_CASE(gradient, *utf::depends_on("warmup")) {
  // Derive the galaxy density
  auto density_g = std::get<0>(law.compute_density(final_density));
  // Derive the gradient, note that ref is absolutely necessary to avoid a SEGV.
  // Otherwise there would be a temporary copy construction in the tuple which would be used in building
  // grad_density_g virtual array. Of course, once the expression completes the temporary is not anymore available.
  // This is not necessary if it is a virtual array as the entire expression is copying down in that case.
  auto grad_density_g = std::get<0>(law.apply_adjoint_gradient(
      final_density, std::make_tuple(std::ref(grad_density))));

  for (size_t i = 0; i < density_g.shape()[0]; i++) {
    for (size_t j = 0; j < density_g.shape()[1]; j++) {
      for (size_t k = 0; k < density_g.shape()[2]; k++) {
        double out = grad_density_g[i][j][k];
        double ref = 2.0 *
                     std::pow(1 + final_density[i][j][k] + EPSILON_VOIDS, 1.0) *
                     grad_density[i][j][k];

        BOOST_CHECK_CLOSE(out, ref, 1e-6);
      }
    }
  }
}


int main(int argc, char **argv)
{
  LibLSS::QUIET_CONSOLE_START = true;
  LibLSS::setupMPI(argc, argv);
  LibLSS::StaticInit::execute();

  int ret = utf::unit_test_main(&init_unit_test, argc, argv);

  LibLSS::StaticInit::finalize();
  LibLSS::doneMPI();

  return ret;
}
