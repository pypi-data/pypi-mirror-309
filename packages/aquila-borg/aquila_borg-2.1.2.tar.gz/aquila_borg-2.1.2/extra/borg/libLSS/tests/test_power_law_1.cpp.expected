Running 3 test cases...
